src/execution/execute-afm.ts

```diff
@@ -1,137 +1,251 @@
-        invariant(dimensionality <= 2, 'executeAfm does not support more than 2 dimensions');
-
-        return this.xhr.post(`/gdc/app/projects/${projectId}/executeAfm`, { body: JSON.stringify(execution) })
-            .then((r => r.getData()))
-            .then((executionResponseWrapper: Execution.IExecutionResponseWrapper) => {
-                const offset = Array(dimensionality).fill(0); // offset holds information on dimensionality
-                const pollingUri = executionResponseWrapper.executionResponse.links.executionResult;
-                return this.getOnePage(pollingUri, offset)
-                    .then((executionResultWrapper: Execution.IExecutionResultWrapper) => {
-                        const responses: Execution.IExecutionResponses = {
-                            executionResponse: executionResponseWrapper.executionResponse,
-                            executionResult: executionResultWrapper ? executionResultWrapper.executionResult : null
-                        };
-                        return responses;
+    public executeAfm(projectId: string, execution: AFM.IExecution)
+        : Promise<Execution.IExecutionResponses> {
+        validateNumOfDimensions(get(execution, 'execution.resultSpec.dimensions').length);
+        return this.getExecutionResponse(projectId, execution)
+            .then((executionResponse: Execution.IExecutionResponse) => {
+                return this.getWholeExecutionResult(executionResponse.links.executionResult)
+                    .then((executionResult: Execution.IExecutionResult | null) => {
+                        return { executionResponse, executionResult };
                     });
             });
     }
 
-    private fetchExecutionResult(pollingUri: string, offset: number[]) {
-        const [uriPart, queryPart] = pollingUri.split(/\?(.+)/);
-        const query = {
-            ...qs.parse(queryPart),
-            limit: getLimit(offset).join(','),
-            offset: offset.join(',')
-        };
-        const finalPollingUri = uriPart + qs.stringify(query, { addQueryPrefix: true });
-        return this.xhr.ajax(finalPollingUri, { method: 'GET' }).then((r) => {
-            if (r.response.status === 204) {
-                return null;
-            }
-            return r.getData();
-        });
+    /**
+     * Get Response from Execution
+     *
+     * @method getExecutionResponse
+     * @param {string} projectId - GD project identifier
+     * @param {AFM.IExecution} execution - See https://github.com/gooddata/gooddata-typings/blob/master/index.ts#L4
+     *
+     * @returns {Promise<Execution.IExecutionResponse>} Promise with `executionResponse`
+     *  See https://github.com/gooddata/gooddata-typings/blob/master/src/Execution.ts#L69
+     */
+    private getExecutionResponse(projectId: string, execution: AFM.IExecution)
+        : Promise<Execution.IExecutionResponse> {
+        validateNumOfDimensions(get(execution, 'execution.resultSpec.dimensions').length);
```

Comment:
I would move this `get(execution, 'execution.resultSpec.dimensions').length` into `validateNumOfDimensions` :)